Tasks/NuGetInstaller/vsonugethelper.ts

```diff
@@ -0,0 +1,184 @@
+﻿/// <reference path="../../definitions/node.d.ts"/>
+/// <reference path="../../definitions/Q.d.ts" />
+/// <reference path="../../definitions/vsts-task-lib.d.ts" />
+
+import fs = require('fs');
+import https = require('https');
+import path = require('path');
+import Q = require('q');
+import tl = require('vsts-task-lib/task');
+import url = require('url');
+
+var xmlreader = require('xmlreader');
+
+interface IPackageSource {
+    id: number;
+    feedName: string;
+    feedUri: string;
+    addCredential: boolean
+}
+
+export class VsoNuGetHelper {
+
+    private _nugetPath: string;
+    private _nugetConfigPath: string;
+    private _accessToken: string;
+
+    private tempNugetConfigDir = path.join(tl.getVariable('agent.buildDirectory'), 'Nuget');
+    public tempNugetConfigPath = path.join(this.tempNugetConfigDir, 'tempNuGet.config');
+  
+    constructor(nugetPath: string, nugetConfigPath: string, accessToken: string) {
+        this._nugetPath = nugetPath;
+        this._nugetConfigPath = nugetConfigPath;
+        this._accessToken = accessToken;
+    }
+
+    public setCredentialsNuGetConfigAndSaveTemp(): Q.Promise<boolean> {
+        var def = Q.defer<boolean>();
+
+        // load content of the user's nuget.config
+        var xml = fs.readFileSync(this._nugetConfigPath).toString();
+
+        // get package sources
+        xmlreader.read(xml, (err, res) => {
+            if (err) return tl.error(err);
+
+            var packageSources = [];
+            var packageSource: IPackageSource;
+            var sourceKey;
+            var sourceValue;
+            var promises: Q.Promise<{}>[] = [];
+            var deffered: Q.Deferred<{}>;
+            var deffers: Q.Deferred<{}>[] = [];
+
+            for (var i = 0; i < res.configuration.packageSources.add.count(); i++) {
+
+                sourceKey = res.configuration.packageSources.add.at(i).attributes().key;
+                sourceValue = res.configuration.packageSources.add.at(i).attributes().value;
+                packageSource = { id: i, feedName: sourceKey, feedUri: sourceValue, addCredential: false };
+                deffered = Q.defer();
+                deffers[i] = deffered;
+                promises.push(deffered.promise);
+
+                // check if need to add credential to feed
+                tl._writeLine('check credential: ' + sourceValue)
+                this.shouldGetCredentialsForFeed(packageSource).then((result) => {                         
+                    if (result.addCredential) {
```

Comment:
async call inside synchronous for loop modifying common data structure, can result in scenarios where data expected in packageSources is not accurate?